wal/src/message_queue_impl/namespace.rs

```diff
@@ -0,0 +1,714 @@
+// Also implement `Debug` for `MessageQueue`.
+impl<M: MessageQueue> fmt::Debug for Namespace<M> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Namespace")
+            .field("namespace", &self.inner.namespace)
+            .field("message_queue", &self.inner.message_queue)
+            .field("meta_encoding", &self.inner.meta_encoding)
+            .field("log_encoding", &self.inner.log_encoding)
+            .finish()
+    }
+}
+
+struct NamespaceInner<M: MessageQueue> {
+    namespace: String,
+    regions: Arc<RwLock<HashMap<RegionId, RegionRef<M>>>>,
+    message_queue: Arc<M>,
+    meta_encoding: MetaEncoding,
+    log_encoding: LogEncoding,
+}
+
+impl<M: MessageQueue> NamespaceInner<M> {
+    pub fn new(namespace: String, message_queue: Arc<M>) -> Self {
+        Self {
+            namespace,
+            regions: Arc::new(RwLock::new(HashMap::new())),
+            message_queue,
+            meta_encoding: MetaEncoding::newest(),
+            log_encoding: LogEncoding::newest(),
+        }
+    }
+
+    /// Get or open the region of `region_id`.
+    ///
+    /// NOTICE: If the region doesn't exist yet, it will be created no matter
+    /// the high-level caller is `write`, `scan`, `read_batch` or
+    /// `mark_delete_entries_up_to` . So, we should consider situations
+    /// about above operations on an empty region.
+    async fn get_or_open_region(
+        &self,
+        region_id: RegionId,
+    ) -> std::result::Result<RegionRef<M>, region::Error> {
+        {
+            let regions = self.regions.read().await;
+            if let Some(region) = regions.get(&region_id) {
+                debug!(
+                    "Region exists and return it, namespace:{}, region id:{}",
+                    self.namespace, region_id
+                );
+                return Ok(region.clone());
+            }
+        }
+
+        let mut regions = self.regions.write().await;
+        // Multiple tables share one region, so double check here is needed.
+        if let Some(region) = regions.get(&region_id) {
+            debug!(
+                "Region exists and return it, namespace:{}, region id:{}",
+                self.namespace, region_id
+            );
+            return Ok(region.clone());
+        }
+
+        let region =
+            Arc::new(Region::open(&self.namespace, region_id, self.message_queue.clone()).await?);
+        regions.insert(region_id, region.clone());
+
+        info!(
+            "Region open successfully, namespace:{}, region id:{}",
+            self.namespace, region_id
+        );
+
+        Ok(region)
+    }
+
+    pub async fn sequence_num(&self, location: Location) -> Result<SequenceNumber> {
+        let region = self
+            .get_or_open_region(location.shard_id as u64)
+            .await
+            .context(GetSequence {
+                namespace: self.namespace.clone(),
+                location,
+            })?;
+
+        Ok(region
```

Comment:
It's a little weird to wrap so many logic inside `Ok(..)`, could you try simplify it?